DRIVERS-3470 case and diacritic text explicit encryption tests (#1931)
mongodb/specifications: source/client-side-encryption/tests/README.md

Reviewer: Suggest renaming this client to `explicitEncryptedClient` to clarify the expected use, and better distinguish it from the client used for auto encryption.

Read the `"_id"` field of `key1Document` as `key1ID`.

Load the file

> [key2-document.json](https://github.com/mongodb/specifications/tree/master/source/client-side-encryption/etc/data/keys/key2-document.json)

Reviewer: `key2-document` is referenced in other tests (e.g. [fle2v2-Compact](https://github.com/mdb-ad/specifications/blob/3166090fe66bf01b01ffa488dcf1491a34bd7241/source/client-side-encryption/etc/test-templates/fle2v2-Compact.yml.template#L11)). Regenerating those test files might have unintended changes.

Rather than have the substring test use two fields in the same schema (which requires two keys due to QE needing a unique key-per-field), I suggest following the pattern with prefix-suffix. Add another schema: `encryptedFields-substring-ci-di.json` to test the insensitive cases.

}
```

> Create a MongoClient with auto-encryption enabled (without `bypassQueryAnalysis`) using these `AutoEncryptionOpts`:

Reviewer: ```suggestion
Create a MongoClient with auto-encryption enabled (without `bypassQueryAnalysis`) named `autoEncryptedClient` using these `AutoEncryptionOpts`:
```

Suggest naming this client `autoEncryptedClient` to distinguish this from client used for explicit encryption.

class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",

>    queryType: "prefix",

Reviewer: Suggest reverting changes of "prefixPreview" => "prefix" and "suffixPreview" => "suffix". I expect these are intended as part of DRIVERS-3321. And when this change is applied later, the tests will need to be updated to require testing 9.0.0+.

Expect an error from libmongocrypt with a message containing the string: "contention factor is required for textPreview
algorithm".

> #### Case 8: can find a case and diacritic-insensitively indexed document by prefix and suffix

Reviewer: ```suggestion
#### Case 8: can find an auto-encrypted case and diacritic-insensitively indexed document by prefix and suffix
```

Suggest using "auto-encrypted" in the test cases to explain why the insert uses auto encryption. This differs from the test cases before, that use documents inserted with explicit-encryption. But auto-encrypted is necessary for the regression test.

> #### Case 8: can find a case and diacritic-insensitively indexed document by prefix and suffix

Reviewer: ```suggestion
#### Case 8: can find a case-insensitively indexed document by prefix and suffix
```

Suggest renaming this test case since the inserted value "BingQiLin" has no diacritics. This wording seems more consistent with Case 9.

> Skip this test case if testing MongoDB server 9.0.0+. This test requires libmongocrypt 1.18.1+.

Reviewer: ```suggestion
This is a regression test for [DRIVERS-3470](https://jira.mongodb.org/browse/DRIVERS-3470). Skip this test case if testing MongoDB server 9.0.0+. This test requires libmongocrypt 1.18.1+.
```

      prefix: PrefixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `explicitEncryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following
filter:

```javascript
{ $expr: { $encStrStartsWith: {input: '$encryptedText', prefix: <encrypted 'bing'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "BingQiLin" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"lin"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "suffix",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: false,
      suffix: SuffixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `explicitEncryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following
filter:

```javascript
{ $expr: { $encStrEndsWith: {input: '$encryptedText', suffix: <encrypted 'lin'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "BingQiLin" }
```

> #### Case 9: can find a diacritic-insensitively indexed document by prefix and suffix

Reviewer: ```suggestion
#### Case 9: can find an auto-encrypted diacritic-insensitively indexed document by prefix and suffix
```

> Skip this test case if testing MongoDB server 9.0.0+. This test requires libmongocrypt 1.18.1+.

Reviewer: ```suggestion
This is a regression test for [DRIVERS-3470](https://jira.mongodb.org/browse/DRIVERS-3470). Skip this test case if testing MongoDB server 9.0.0+. This test requires libmongocrypt 1.18.1+.
```

      prefix: PrefixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `explicitEncryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following
filter:

```javascript
{ $expr: { $encStrStartsWith: {input: '$encryptedText', prefix: <encrypted 'cafe'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "cafébarbäz" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"baz"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "suffix",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: false,
      suffix: SuffixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `explicitEncryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following
filter:

```javascript
{ $expr: { $encStrEndsWith: {input: '$encryptedText', suffix: <encrypted 'baz'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "cafébarbäz" }
```

> #### Case 10: can find a case-insensitively indexed document by substring

Reviewer: ```suggestion
#### Case 10: can find an auto-encrypted case-insensitively indexed document by substring
```

> This test requires libmongocrypt 1.18.1+.

Reviewer: ```suggestion
This is a regression test for [DRIVERS-3470](https://jira.mongodb.org/browse/DRIVERS-3470). This test requires libmongocrypt 1.18.1+.
```

Use `autoEncryptedClient` to insert the following document into `db.substring-ci-di` with majority write concern:

```javascript
{ "encryptedText": "FooBarBaz" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"bar"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",

>    queryType: "substring",

Reviewer: ```suggestion
   queryType: "substringPreview",
```

   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: false,
      substring: SubstringOpts {
        strMaxLength: 10,
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
      }
   },
}
```

Use `explicitEncryptedClient` to run a "find" operation on the `db.substring-ci-di` collection with the following
filter:

```javascript
{ $expr: { $encStrContains: {input: '$encryptedText', substring: <encrypted 'bar'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "FooBarBaz" }
```

> #### Case 11: can find a diacritic-insensitively indexed document by substring

Reviewer: ```suggestion
#### Case 11: can find an auto-encrypted diacritic-insensitively indexed document by substring
```

> This test requires libmongocrypt 1.18.1+.

Reviewer: ```suggestion
This is a regression test for [DRIVERS-3470](https://jira.mongodb.org/browse/DRIVERS-3470). This test requires libmongocrypt 1.18.1+.
```

Use `autoEncryptedClient` to insert the following document into `db.substring-ci-di` with majority write concern:

```javascript
{ "encryptedText": "foocafébaz" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"cafe"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",

>    queryType: "substring",

Reviewer: ```suggestion
   queryType: "substringPreview",
```

Expect an error from libmongocrypt with a message containing the string: "contention factor is required for textPreview
algorithm".

> #### Case 8: can find a case and diacritic-insensitively indexed document by prefix and suffix

Reviewer: Also note: "Skip this test case if testing MongoDB server 9.0.0+." note (similar to Case 7) to cases 8 and 9 (requiring prefixPreview and suffixPreview).

And note that libmongocrypt 1.18.1 is required for test cases 8 and 9 since those are regression tests for [MONGOCRYPT-903](https://jira.mongodb.org/browse/MONGOCRYPT-903).

Reviewer 2: +1

`caseSensitive: false, diacriticSensitive: false` suffix configuration of `db.prefix-suffix-ci-di`.

Use `clientEncryption.encrypt()` to encrypt the string `"bing"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "prefix",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: false,
      prefix: PrefixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrStartsWith: {input: '$encryptedText', prefix: <encrypted 'bing'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "BingQiLin" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"lin"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "suffix",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,

>       diacriticSensitive: true,

Reviewer 2: The suffix `EncryptOpts` here use `diacriticSensitive: true`, but `db.prefix-suffix-ci-di` is indexed with `diacriticSensitive: false`. Could you clarify if this is intentional?

Author: Fixed, they should all match

        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrEndsWith: {input: '$encryptedText', suffix: <encrypted 'lin'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "BingQiLin" }
```

#### Case 9: can find a diacritic-insensitively indexed document by prefix and suffix

Create a MongoClient with auto-encryption enabled (without `bypassQueryAnalysis`) using these `AutoEncryptionOpts`:

```typescript
class AutoEncryptionOpts {
   keyVaultNamespace: "keyvault.datakeys",
   kmsProviders: { "local": { "key": <base64 decoding of LOCAL_MASTERKEY> } },
}
```

Use that client to insert the following document into `db.prefix-suffix-ci-di` with majority write concern:

```javascript
{ "encryptedText": "cafébarbäz" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"cafe"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "prefix",
   contentionFactor: 0,
   textOpts: TextOpts {

>       caseSensitive: true,

Reviewer 2: Same concern here: `caseSensitive: true` is used in both encrypt opts, but the collection is indexed with `caseSensitive: false`. The query strings "cafe" and "baz" differ from the stored "cafébarbäz" only in diacritics, not case — so `caseSensitive: true` may not affect the result in practice. Is this intentional?

Expect an error from libmongocrypt with a message containing the string: "contention factor is required for textPreview
algorithm".

#### Case 8: can find a case and diacritic-insensitively indexed document by prefix and suffix

Skip this test case if testing MongoDB server 9.0.0+. This test requires libmongocrypt 1.18.1+.

> `"BingQiLin"` was inserted via auto-encryption in Test Setup, indexed under the

Reviewer: Case 9, 10, 11, and 12 have the insert step within the test case. Suggest doing the same in Case 8 for consistency. Move the insert step from the test setup into this case:

```markdown
Use `autoEncryptedClient` to insert the following document into `db.prefix-suffix-ci-di` with majority write concern: [...]
```

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "suffix",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: false,
      suffix: SuffixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrEndsWith: {input: '$encryptedText', suffix: <encrypted 'lin'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "BingQiLin" }
```

#### Case 9: can find a diacritic-insensitively indexed document by prefix and suffix

Skip this test case if testing MongoDB server 9.0.0+. This test requires libmongocrypt 1.18.1+.

Create a MongoClient with auto-encryption enabled (without `bypassQueryAnalysis`) using these `AutoEncryptionOpts`:

```typescript
class AutoEncryptionOpts {
   keyVaultNamespace: "keyvault.datakeys",
   kmsProviders: { "local": { "key": <base64 decoding of LOCAL_MASTERKEY> } },
}
```

> Use that client to insert the following document into `db.prefix-suffix-ci-di` with majority write concern:

Reviewer: Suggest reusing the client created in set-up:

```suggestion
Use `autoEncryptedClient` to insert the following document into `db.prefix-suffix-ci-di` with majority write concern:
```

   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrStartsWith: {input: '$encryptedText', prefix: <encrypted 'cafe'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "cafébarbäz" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"baz"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "suffix",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: false,
      suffix: SuffixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrEndsWith: {input: '$encryptedText', suffix: <encrypted 'baz'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "cafébarbäz" }
```

#### Case 10: can find a case-insensitively indexed document by substring

This test requires libmongocrypt 1.18.1+.

> Use `clientEncryption.encrypt()` to encrypt the string `"FooBarBaz"` with the following `EncryptOpts`:

Reviewer: I expect the inserted value needs to use auto encryption to serve as a regression test. [MONGOCRYPT-903](https://jira.mongodb.org/browse/MONGOCRYPT-903) flipped the booleans, so I expect explicit encryption for both insert and find would still succeed without the bug fix.

Suggest changing this step to:

```markdown
Use `autoEncryptedClient` to insert the following document into `db.substring-ci-di` with majority write concern: [...]
```

A similar suggestion applies to case 11.

Expect an error from libmongocrypt with a message containing the string: "contention factor is required for textPreview
algorithm".

#### Case 8: can find a case and diacritic-insensitively indexed document by prefix and suffix

> Skip this test case if testing MongoDB server 9.0.0+.

Reviewer: ```suggestion
Skip this test case if testing MongoDB server 9.0.0+. This test requires libmongocrypt 1.18.1+.
```

     }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrStartsWith: {input: '$encryptedText', prefix: <encrypted 'bing'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "BingQiLin" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"lin"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "suffix",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: false,
      suffix: SuffixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrEndsWith: {input: '$encryptedText', suffix: <encrypted 'lin'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "BingQiLin" }
```

#### Case 9: can find a diacritic-insensitively indexed document by prefix and suffix

> Skip this test case if testing MongoDB server 9.0.0+.

Reviewer: ```suggestion
Skip this test case if testing MongoDB server 9.0.0+. This test requires libmongocrypt 1.18.1+.
```

      prefix: PrefixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrStartsWith: {input: '$encryptedText', prefix: <encrypted 'cafe'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "cafébarbäz" }
```

Use `clientEncryption.encrypt()` to encrypt the string `"baz"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key1ID>,
   algorithm: "TextPreview",
   queryType: "suffix",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: false,
      suffix: SuffixOpts {
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
     }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.prefix-suffix-ci-di` collection with the following filter:

```javascript
{ $expr: { $encStrEndsWith: {input: '$encryptedText', suffix: <encrypted 'baz'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedText": "cafébarbäz" }
```

> #### Case 10: can find a case-insensitively indexed document by substring

Reviewer: ```suggestion
#### Case 10: can find a case-insensitively indexed document by substring

This test requires libmongocrypt 1.18.1+.
```

   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: true,
      substring: SubstringOpts {
        strMaxLength: 10,
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
      },
   },
}
```

Use `encryptedClient` to insert the following document into `db.substring` with majority write concern:

```javascript
{ "encryptedTextInsensitive": <encrypted 'FooBarBaz'> }
```

Use `clientEncryption.encrypt()` to encrypt the string `"bar"` with the following `EncryptOpts`:

```typescript
class EncryptOpts {
   keyId : <key2ID>,
   algorithm: "TextPreview",
   queryType: "substring",
   contentionFactor: 0,
   textOpts: TextOpts {
      caseSensitive: false,
      diacriticSensitive: true,
      substring: SubstringOpts {
        strMaxLength: 10,
        strMaxQueryLength: 10,
        strMinQueryLength: 2,
      }
   },
}
```

Use `encryptedClient` to run a "find" operation on the `db.substring` collection with the following filter:

```javascript
{ $expr: { $encStrContains: {input: '$encryptedTextInsensitive', substring: <encrypted 'bar'>} } }
```

Assert the following document is returned:

```javascript
{ "encryptedTextInsensitive": "FooBarBaz" }
```

> #### Case 11: can find a diacritic-insensitively indexed document by substring

Reviewer: ```suggestion
#### Case 11: can find a diacritic-insensitively indexed document by substring

This test requires libmongocrypt 1.18.1+.
```
